File: ruby/test/web_socket_transport_test.rb
# frozen_string_literal: true

require "test_helper"

class WebSocketTransportTest < CableTest
  Opcode = ActionCableClient::WebSocketTransport::Opcode
  Protocol = ActionCableClient::Protocol

  def test_web_socket_transport_negotiates_the_subprotocol
    server = loopback_server
    connection = dial(server, subprotocols: [ Protocol::V1JSON::SUBPROTOCOL ])

    assert_equal Protocol::V1JSON::SUBPROTOCOL, connection.subprotocol
    assert_equal Protocol::V1JSON::SUBPROTOCOL, server.accept.header("Sec-WebSocket-Protocol"),
      "expected the client to offer the subprotocol"
  end

  def test_web_socket_transport_sends_headers
    server = loopback_server
    dial(server, subprotocols: [ Protocol::V1JSON::SUBPROTOCOL ],
      headers: { "Cookie" => "session=secret", "Origin" => "https://example.com" })

    peer = server.accept
    assert_equal "session=secret", peer.header("Cookie")
    assert_equal "https://example.com", peer.header("Origin")
    assert_equal ActionCableClient::WebSocketTransport::USER_AGENT, peer.header("User-Agent")
    assert_equal "/cable", peer.request.target
  end

  def test_web_socket_transport_sends_the_callers_user_agent
    server = loopback_server
    dial(server, headers: { "User-Agent" => "custom-agent" })

    assert_equal "custom-agent", server.accept.header("User-Agent")
  end

  def test_web_socket_transport_neutralizes_header_injection
    server = loopback_server
    dial(server, headers: { "Authorization" => "Bearer token\r\nX-Injected: gotcha" })

    peer = server.accept
    assert_nil peer.header("X-Injected"), "expected the newlines to be neutralized"
    assert peer.header("Authorization").start_with?("Bearer token"),
      "expected the authorization header to survive, got #{peer.header("Authorization").inspect}"
  end

  def test_web_socket_transport_round_trips_messages
    server = loopback_server
    connection = dial(server, subprotocols: [ Protocol::V1JSON::SUBPROTOCOL ])
    peer = server.accept

    connection.write(%({"command":"subscribe"}))
    assert_equal %({"command":"subscribe"}), peer.read

    peer.write(Opcode::TEXT, %({"type":"welcome"}))
    assert_equal %({"type":"welcome"}), connection.read(timeout: WAIT)
  end

  def test_web_socket_transport_answers_pings
    server = loopback_server
    connection = dial(server)
    peer = server.accept

    peer.write(Opcode::PING, "beat")
    peer.write(Opcode::TEXT, "after the ping")

    assert_equal "after the ping", connection.read(timeout: WAIT)

    frame = peer.read_frame
    assert_equal Opcode::PONG, frame.opcode, "expected a pong"
    assert_equal "beat", frame.payload, "expected the pong to carry the ping payload"
  end

  def test_web_socket_transport_reassembles_fragments
    server = loopback_server
    connection = dial(server)
    peer = server.accept

    peer.write_fragment(Opcode::TEXT, "one ", final: false)
    peer.write_fragment(Opcode::PING, "interleaved", final: true)
    peer.write_fragment(Opcode::CONTINUATION, "message", final: true)

    assert_equal "one message", connection.read(timeout: WAIT)
  end

  def test_web_socket_transport_reads_large_messages
    server = loopback_server
    connection = dial(server)
    peer = server.accept

    long = "cable" * 30_000
    peer.write(Opcode::TEXT, long)
    assert_equal long, connection.read(timeout: WAIT)

    connection.write(long)
    assert_equal long, peer.read
  end

  def test_web_socket_transport_refuses_oversized_messages
    server = loopback_server
    connection = dial(server, max_message_size: 8)

    server.accept.write(Opcode::TEXT, "far too long for eight bytes")

    assert_raises(ActionCableClient::MessageTooBigError) { connection.read(timeout: WAIT) }
  end

  def test_web_socket_transport_refuses_oversized_fragmented_messages
    server = loopback_server
    connection = dial(server, max_message_size: 8)

    peer = server.accept
    peer.write_fragment(Opcode::TEXT, "five ", final: false)
    peer.write_fragment(Opcode::CONTINUATION, "more", final: true)

    assert_raises(ActionCableClient::MessageTooBigError) { connection.read(timeout: WAIT) }
  end

  def test_web_socket_transport_reports_server_close
    server = loopback_server
    connection = dial(server)

    server.accept.write(Opcode::CLOSE, [ 4401 ].pack("n") + "unauthorized")

    error = assert_raises(ActionCableClient::CloseError) { connection.read(timeout: WAIT) }
    assert_equal 4401, error.code
    assert_equal "unauthorized", error.reason
  end

  def test_web_socket_transport_reports_a_server_close_without_a_status
    server = loopback_server
    connection = dial(server)

    server.accept.write(Opcode::CLOSE)

    error = assert_raises(ActionCableClient::CloseError) { connection.read(timeout: WAIT) }
    assert_equal 1005, error.code
    assert_equal "", error.reason
  end

  def test_web_socket_transport_closes_with_a_status
    server = loopback_server
    connection = dial(server)
    peer = server.accept

    assert_respond_to connection, :close_with_status, "the built-in connection should be able to say why"
    connection.close_with_status(4000, "done here")

    frame = peer.read_frame
    assert_equal Opcode::CLOSE, frame.opcode
    assert_equal 4000, frame.payload.unpack1("n")
    assert_equal "done here", frame.payload.byteslice(2..)
  end

  def test_web_socket_transport_truncates_a_close_reason_to_fit_the_frame
    server = loopback_server
    connection = dial(server)
    peer = server.accept

    connection.close_with_status(4000, "r" * 200)

    frame = peer.read_frame
    assert_equal Opcode::CLOSE, frame.opcode
    assert_equal 125, frame.payload.bytesize, "a control frame's payload is at most 125 bytes"
  end

  def test_web_socket_transport_refuses_a_non_upgrade_response
    server = loopback_server
    server.refusal = "HTTP/1.1 404 Not Found\r\nContent-Length: 13\r\n\r\nno cable here"

    error = assert_raises(ActionCableClient::HandshakeError) { transport.dial(server.url) }
    assert_equal 404, error.status_code
    assert_equal "404 Not Found", error.status
  end

  def test_web_socket_transport_does_not_follow_a_redirect
    server = loopback_server
    server.refusal = "HTTP/1.1 302 Found\r\nLocation: /elsewhere\r\nContent-Length: 0\r\n\r\n"

    error = assert_raises(ActionCableClient::HandshakeError) { transport.dial(server.url) }
    assert_equal 302, error.status_code
  end

  def test_web_socket_transport_refuses_a_bad_accept_key
    server = loopback_server
    server.bad_accept = true

    assert_raises(ActionCableClient::Error) { transport.dial(server.url) }
  end

  def test_web_socket_transport_honors_the_read_timeout
    server = loopback_server
    connection = dial(server)
    server.accept

    assert_raises(ActionCableClient::TimeoutError) { connection.read(timeout: 0.05) }
  end

  def test_web_socket_transport_refuses_a_masked_server_frame
    server = loopback_server
    connection = dial(server)

    # RFC 6455 §5.1: a server must never mask, and a client that sees a masked
    # frame must fail the connection rather than quietly unmask it.
    server.accept.write_masked(Opcode::TEXT, %({"type":"welcome"}))

    assert_raises(ActionCableClient::Error) { connection.read(timeout: WAIT) }
  end

  def test_web_socket_transport_replies_to_a_close_once
    server = loopback_server
    connection = dial(server)
    peer = server.accept

    peer.write(Opcode::CLOSE, [ 1000 ].pack("n"))
    assert_raises(ActionCableClient::CloseError) { connection.read(timeout: WAIT) }
    connection.close

    assert_equal 1, peer.close_frames(timeout: 0.2), "expected exactly one close frame in reply"
  end

  # Runs the whole cable dance over an actual WebSocket connection.
  def test_client_over_the_real_transport
    server = loopback_server
    client = ActionCableClient.new(server.url, logger: test_logger)
    (@clients ||= []) << client

    connected = connecting(client)
    peer = server.accept
    peer.write(Opcode::TEXT, %({"type":"welcome"}))
    connected.value

    pending = subscribing(client)
    assert_equal %({"command":"subscribe","identifier":"{\\"channel\\":\\"RoomChannel\\",\\"id\\":42}"}), peer.read
    peer.write(Opcode::TEXT, %({"type":"confirm_subscription","identifier":"{\\"channel\\":\\"RoomChannel\\",\\"id\\":42}"}))
    subscription = pending.value

    peer.write(Opcode::TEXT,
      %({"identifier":"{\\"channel\\":\\"RoomChannel\\",\\"id\\":42}","message":{"body":"Hello!"}}))
    assert_equal %({"body":"Hello!"}), subscription.next_message(timeout: WAIT).to_s

    subscription.perform("speak", body: "Hi!")
    assert_equal %({"command":"message","identifier":"{\\"channel\\":\\"RoomChannel\\",\\"id\\":42}",) +
      %("data":"{\\"action\\":\\"speak\\",\\"body\\":\\"Hi!\\"}"}), peer.read
  end

  def teardown
    @connections&.each(&:close)
    super
  end

  private
    def transport(**options)
      ActionCableClient::WebSocketTransport.new(**options)
    end

    def dial(server, subprotocols: [], headers: {}, **options)
      connection = transport(**options).dial(server.url, subprotocols: subprotocols, headers: headers)
      (@connections ||= []) << connection

      connection
    end
end

File: kotlin/client/src/jvmTest/kotlin/com/basecamp/actioncable/WebSocketTransportTest.kt
package com.basecamp.actioncable

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import kotlin.coroutines.cancellation.CancellationException
import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertFails
import kotlin.test.assertFailsWith
import kotlin.test.assertIs
import kotlin.test.assertNull
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds

/**
 * The built-in transport against a real loopback server that does the upgrade
 * by hand. Everything here goes over a socket, which is the point: it is what
 * proves the WebSocket code rather than the client's idea of it.
 */
class WebSocketTransportTest {
    @Test
    fun `negotiates the subprotocol`() = transportTest { server ->
        val conn = dial(server, DialOptions(subprotocols = listOf(SUBPROTOCOL_V1_JSON)))

        assertEquals(SUBPROTOCOL_V1_JSON, conn.subprotocol)
        assertEquals(
            SUBPROTOCOL_V1_JSON,
            server.accept().request.header("Sec-WebSocket-Protocol"),
            "expected the client to offer the subprotocol",
        )
    }

    @Test
    fun `sends headers`() = transportTest { server ->
        dial(
            server,
            DialOptions(
                subprotocols = listOf(SUBPROTOCOL_V1_JSON),
                headers = Headers.of("Cookie" to "session=secret", "Origin" to "https://example.com"),
            ),
        )

        val request = server.accept().request
        assertEquals("session=secret", request.header("Cookie"))
        assertEquals("https://example.com", request.header("Origin"))
        assertEquals(WebSocketTransport.USER_AGENT, request.header("User-Agent"))
        assertEquals("/cable", request.path)
    }

    @Test
    fun `sends the caller's user agent`() = transportTest { server ->
        dial(server, DialOptions(headers = Headers.of("User-Agent" to "custom-agent")))

        assertEquals("custom-agent", server.accept().request.header("User-Agent"))
    }

    @Test
    fun `neutralizes header injection`() = transportTest { server ->
        dial(server, DialOptions(headers = Headers.of("Authorization" to "Bearer token\r\nX-Injected: gotcha")))

        val request = server.accept().request
        assertNull(request.header("X-Injected"), "expected the newlines to be neutralized")
        assertContains(
            request.header("Authorization").orEmpty(),
            "Bearer token",
            message = "expected the authorization header to survive",
        )
    }

    @Test
    fun `round trips messages`() = transportTest { server ->
        val conn = dial(server, DialOptions(subprotocols = listOf(SUBPROTOCOL_V1_JSON)))
        val peer = server.accept()

        conn.write("""{"command":"subscribe"}""".encodeToByteArray())
        assertEquals("""{"command":"subscribe"}""", peer.read())

        peer.write(OP_TEXT, """{"type":"welcome"}""")
        assertEquals("""{"type":"welcome"}""", conn.readWithin().decodeToString())
    }

    @Test
    fun `answers pings`() = transportTest { server ->
        val conn = dial(server, DialOptions())
        val peer = server.accept()

        peer.write(OP_PING, "beat")
        peer.write(OP_TEXT, "after the ping")

        assertEquals("after the ping", conn.readWithin().decodeToString())

        val frame = peer.readFrame()
        assertEquals(OP_PONG, frame.opcode, "expected a pong")
        assertEquals("beat", frame.text, "expected the pong to carry the ping payload")
    }

    @Test
    fun `reassembles fragments`() = transportTest { server ->
        val conn = dial(server, DialOptions())
        val peer = server.accept()

        peer.writeFragment(OP_TEXT, "one ".encodeToByteArray(), false)
        peer.writeFragment(OP_PING, "interleaved".encodeToByteArray(), true)
        peer.writeFragment(OP_CONTINUATION, "message".encodeToByteArray(), true)

        assertEquals("one message", conn.readWithin().decodeToString())
    }

    @Test
    fun `reads large messages`() = transportTest { server ->
        val conn = dial(server, DialOptions())
        val peer = server.accept()

        val long = "cable".repeat(30_000)
        peer.write(OP_TEXT, long)
        assertEquals(long, conn.readWithin().decodeToString())

        conn.write(long.encodeToByteArray())
        assertEquals(long, peer.read())
    }

    @Test
    fun `refuses oversized messages`() = transportTest { server ->
        val conn = WebSocketTransport(maxMessageSize = 8).dial(server.url, DialOptions())

        server.accept().write(OP_TEXT, "far too long for eight bytes")

        assertFailsWith<MessageTooBigException> { conn.readWithin() }
    }

    @Test
    fun `refuses oversized fragmented messages`() = transportTest { server ->
        val conn = WebSocketTransport(maxMessageSize = 8).dial(server.url, DialOptions())

        val peer = server.accept()
        peer.writeFragment(OP_TEXT, "five ".encodeToByteArray(), false)
        peer.writeFragment(OP_CONTINUATION, "more".encodeToByteArray(), true)

        assertFailsWith<MessageTooBigException> { conn.readWithin() }
    }

    @Test
    fun `reports server close`() = transportTest { server ->
        val conn = dial(server, DialOptions())

        server.accept().write(OP_CLOSE, closePayload(4401, "unauthorized"))

        val closed = assertFailsWith<CloseException> { conn.readWithin() }
        assertEquals(4401, closed.code)
        assertEquals("unauthorized", closed.reason)
    }

    @Test
    fun `reports a server close without a status`() = transportTest { server ->
        val conn = dial(server, DialOptions())

        server.accept().write(OP_CLOSE, ByteArray(0))

        val closed = assertFailsWith<CloseException> { conn.readWithin() }
        assertEquals(1005, closed.code)
        assertEquals("", closed.reason)
    }

    @Test
    fun `closes with a status`() = transportTest { server ->
        val conn = dial(server, DialOptions())
        val peer = server.accept()

        val closer = assertIs<StatusCloser>(conn, "the built-in connection should implement StatusCloser")
        closer.closeWithStatus(4000, "done here")

        val frame = peer.readFrame()
        assertEquals(OP_CLOSE, frame.opcode)
        assertEquals(4000, statusOf(frame.payload))
        assertEquals("done here", frame.payload.decodeToString(2, frame.payload.size))
    }

    @Test
    fun `truncates a close reason to fit the frame`() = transportTest { server ->
        val conn = dial(server, DialOptions())
        val peer = server.accept()

        (conn as StatusCloser).closeWithStatus(4000, "r".repeat(200))

        val frame = peer.readFrame()
        assertEquals(OP_CLOSE, frame.opcode)
        assertEquals(125, frame.payload.size, "a control frame's payload is at most 125 bytes")
    }

    @Test
    fun `refuses a non-upgrade response`() = runTest(timeout = 30.seconds) {
        withContext(Dispatchers.Default) {
            LoopbackServer(refuseWith = notFound()).use { server ->
                val refused =
                    assertFailsWith<HandshakeException> { WebSocketTransport().dial(server.url, DialOptions()) }
                assertEquals(404, refused.statusCode)
            }
        }
    }

    @Test
    fun `does not follow a redirect`() = runTest(timeout = 30.seconds) {
        withContext(Dispatchers.Default) {
            LoopbackServer(refuseWith = redirect()).use { server ->
                val refused =
                    assertFailsWith<HandshakeException> { WebSocketTransport().dial(server.url, DialOptions()) }
                assertEquals(302, refused.statusCode)
            }
        }
    }

    @Test
    fun `refuses a bad accept key`() = runTest(timeout = 30.seconds) {
        withContext(Dispatchers.Default) {
            LoopbackServer(badAccept = true).use { server ->
                assertFailsWith<ActionCableException> { WebSocketTransport().dial(server.url, DialOptions()) }
            }
        }
    }

    @Test
    fun `honors cancellation`() = transportTest { server ->
        val conn = dial(server, DialOptions())
        server.accept()

        assertFailsWith<CancellationException> {
            withTimeout(50.milliseconds) { conn.read() }
        }
    }

    /** The whole cable dance over an actual WebSocket connection. */
    @Test
    fun `the client over the real transport`() = runTest(timeout = 30.seconds) {
        withContext(Dispatchers.Default) {
            LoopbackServer().use { server ->
                val test = CableTest()
                val client = ActionCableClient(server.url, logger = { println(it) })

                try {
                    val connecting = test.connecting(client)
                    val peer = server.accept()
                    peer.write(OP_TEXT, """{"type":"welcome"}""")
                    connecting.await()

                    val subscribing = test.subscribing(client)
                    assertEquals(
                        """{"command":"subscribe","identifier":"{\"channel\":\"RoomChannel\",\"id\":42}"}""",
                        peer.read(),
                    )
                    peer.write(
                        OP_TEXT,
                        """{"type":"confirm_subscription","identifier":"{\"channel\":\"RoomChannel\",\"id\":42}"}""",
                    )

                    val subscription = subscribing.await()
                    val messages = test.reading(subscription)

                    peer.write(
                        OP_TEXT,
                        """{"identifier":"{\"channel\":\"RoomChannel\",\"id\":42}","message":{"body":"Hello!"}}""",
                    )
                    assertEquals("""{"body":"Hello!"}""", messages.next().toString())

                    subscription.perform("speak", mapOf("body" to "Hi!"))
                    assertEquals(
                        """{"command":"message","identifier":"{\"channel\":\"RoomChannel\",\"id\":42}",""" +
                            """"data":"{\"action\":\"speak\",\"body\":\"Hi!\"}"}""",
                        peer.read(),
                    )
                } finally {
                    client.close()
                    test.tearDown()
                }
            }
        }
    }

    @Test
    fun `refuses a masked server frame`() = transportTest { server ->
        val conn = dial(server, DialOptions())

        // RFC 6455 §5.1: a server must never mask, and a client that sees a
        // masked frame must fail the connection rather than quietly unmask it.
        server.accept().writeMasked(OP_TEXT, """{"type":"welcome"}""".encodeToByteArray())

        assertFails { conn.readWithin() }
    }

    @Test
    fun `replies to a close once`() = transportTest { server ->
        val conn = dial(server, DialOptions())
        val peer = server.accept()

        peer.write(OP_CLOSE, closePayload(1000, ""))
        assertFailsWith<CloseException> { conn.readWithin() }
        conn.close()

        assertEquals(1, peer.closeFrames(), "expected exactly one close frame in reply")
    }
}

/** A transport test with a loopback server up, torn down however the test ends. */
private fun transportTest(body: suspend (LoopbackServer) -> Unit) = runTest(timeout = 30.seconds) {
    withContext(Dispatchers.Default) {
        LoopbackServer().use { body(it) }
    }
}

private suspend fun dial(server: LoopbackServer, options: DialOptions): Conn = WebSocketTransport().dial(server.url, options)

private suspend fun Conn.readWithin(): ByteArray = withTimeout(5.seconds) { read() }

private fun closePayload(code: Int, reason: String): ByteArray =
    byteArrayOf((code shr 8).toByte(), code.toByte()) + reason.encodeToByteArray()

private fun statusOf(payload: ByteArray): Int = (payload[0].toInt() and 0xff) shl 8 or (payload[1].toInt() and 0xff)

private fun notFound(): String =
    "HTTP/1.1 404 Not Found\r\nContent-Type: text/plain\r\nContent-Length: 13\r\nConnection: close\r\n\r\nno cable here"

private fun redirect(): String = "HTTP/1.1 302 Found\r\nLocation: /elsewhere\r\nContent-Length: 0\r\nConnection: close\r\n\r\n"
